First version of connectivity-manager

martin-ejdestig opened the pull request.

TSAnsari commented:
Can this huge change be divided into separate smaller/independent changes? Then it will be easier to review by concentrating on smaller changes. Currently there are 70 files in this single commit.

martin-ejdestig commented:
Could perhaps split it up in multiple commits. But it would probably not be anything other than "first add these files, then these, and then these". :/

I will add some more high level documentation in the start of some files/classes. Should make it easier to get an overview of how it all fits together and how the individual parts work.

But yes, this is a large pr... thousands of lines. More complicated than one would think to connect to an access point with ConnMan by proxying D-Bus calls and have some abstraction in between. Not much to do about that. :(

martin-ejdestig commented:
I should make a better commit message as well. Added to my TODO-list for this pr.

martin-ejdestig commented:
>  But it would probably not be anything other than "first add these files, then these, and then these".

Thinking some more on this... Perhaps it will make it easier to review if I split it up in e.g. "Add D-Bus interface", "Add helper struct for handling credentials", etc. but not split up the daemon parts so much. Will try to do this.

DunderRoffe commented on src/daemon/backends/connman_manager.h:
```diff
@@ -0,0 +1,100 @@
+
+        void name_owner_changed() const;
+
+        void get_technologies_finish(const Glib::RefPtr<Gio::AsyncResult> &result) const;
```
What is the `get` part of this function? There are no return values and no output variables...
Also does finish mean that it is blocking or something else?

martin-ejdestig replied:
The `get_` part comes from the fact that it is the slot/method called when the asynchronous D-Bus method "GetTechnologies" is finished. You will see the connection if you look at where it is used in `connman_manager.cpp`.

DunderRoffe commented on src/daemon/dbus_name_watcher.h:
```diff
@@ -0,0 +1,65 @@
+    // Meant to be used when service should call a method in client (as is the case for e.g. the
+    // UserInputAgent interface). Service needs to know if a client disconnects from the bus before
+    // method should be called.
+    class DBusNameWatcher
```
Seems like this could be useful for other projects as well? Maybe consider moving it to the generator or a utils lib?

martin-ejdestig replied:
Maybe. Unsure what we will do long term with reusable code for our components. I think it makes sense to do stuff for a while and the look at what makes sense to share when we have a clearer picture. (Of course, this relies on us getting the time to refactor.)

DunderRoffe replied:
:+1:

martin-ejdestig commented on src/cli/input_handler.h:
```diff
@@ -0,0 +1,67 @@
+//
+// SPDX-License-Identifier: MPL-2.0
+
+#ifndef CONNECTIVITY_MANAGER_CLI_INPUT_HANDLER_H
```
Not standard.

In practice, it is available everywhere we care about. I actually thought about using it when starting writing this code (and later again when moving some of it out to the D-Bus template) since I know it has been something practiced in previous projects here. But, decided against it since it is mentioned that it should not be used in some guidelines that @JEderonn wanted to use.

martin-ejdestig commented on src/cli/main.cpp:
```diff
@@ -13,48 +13,45 @@
-    if (proxy->dbusProxy()->get_name_owner().empty()) {
-        std::cout << "Service not available, quitting.\n";
+    auto arguments = ConnectivityManager::Cli::Arguments::parse(argc, argv, std::cout);
+    if (!arguments)
```
I agree, `auto` should not be used too much. In this case the line becomes so long though. `std::optional<ConnectivityManager::Cli::Arguments>` is the type.

martin-ejdestig replied:
Using type aliases now (not ns aliases, should be discouraged).

martin-ejdestig commented on src/daemon/backend.cpp:
```diff
@@ -0,0 +1,165 @@
+    {
+        auto i = state_.wifi.access_points.find(access_point.id);
+        if (i == state_.wifi.access_points.cend()) {
+            assert(false);
```
I do not know. I think I would rather just remove the `assert()`. Will see how the meson.bbclass works by default. If it does not make it so `-DNDEBUG` is passed in release builds we should probably not user assert() too much since it should be used to indicate "programming error"... but lead to unexpected crashes in production systems.

martin-ejdestig replied:
Turns out poky does not add -DNDEBUG by default. Removed this assert since it is not critical. Resolving.

DunderRoffe commented on README.md:
```diff
@@ -1,6 +1,6 @@
+# Connectivity Manager
 
-Template for creating a D-Bus service component.
+D-Bus service for managing connectivity. TODO: Write more here.
```
Please specify how to build and what dependencies there are. Does not need to be overly complicated but at least state something like "build using regular Meson setup". I do not know enough about meson to say if such a thing exists anyways... Also state how to run the tests.

martin-ejdestig replied:
Added basic build instructions in template and rebased this pr to that template revision (since it is the "first commit", it is fine to do this).

martin-ejdestig commented on src/daemon/backend.cpp:
```diff
@@ -0,0 +1,165 @@
+        }
+
+        const WiFiAccessPoint copy = std::move(i->second);
+        state_.wifi.access_points.erase(i);
```
`std::unordered_map::erase()` you mean? The iterator is valid (set just above), so no.

martin-ejdestig replied:
"no" as in "no, it can not fail".

martin-ejdestig replied:
Do not see a problem here, resolving. Plx reopen if I misunderstood.

martin-ejdestig commented on src/daemon/backends/connman_agent.cpp:
```diff
@@ -0,0 +1,336 @@
+                    if (credentials.password->value.empty())
+                        credentials.password->value = previous_password->value;
+
+                } else if (credentials.password_alternative &&
```
It looks like it does here since the ConnMan agent API is underspecified... and bad. There is a way to have "values" in "fields". They could have used that for passing any previous password. But instead they added another field for previous password.

Perhaps I do not understand what you mean though.

DunderRoffe replied:
Hmm so I guess I was a bit confused ... It is more clear when I re-read the code and I seem to have misunderstood it .

DunderRoffe replied:
Maybe it could be broken out in to functions that handle different cases to make it easier to read? It is hard to say how this can be done in a more readable way

martin-ejdestig replied:
Perhaps this warrants some documentation at the start of the function... since the D-Bus API is so... wönderfül. I scratched my head over https://git.kernel.org/pub/scm/network/connman/connman.git/tree/doc/agent-api.txt#n97 ... a lot.

martin-ejdestig replied:
Added a short blurb at top of function trying to explain the "Value" vs "PreviousPassphrase" thing and that there are a couple of examples in doc/agent-api.txt . Suggestions for how to improve it further welcome... I think I am to tired to come up with anything better at the moment.

martin-ejdestig replied:
Resolving. Please reopen/comment again if you have any ideas for how to improve it.

martin-ejdestig commented on src/daemon/backends/connman_agent.h:
```diff
@@ -0,0 +1,95 @@
+
+    // The way ConnMan is implemented now, agent_released() will only be called when ConnMan exits
+    // cleanly. Need to clear any pending service connects that may rely on agent being registered.
+    class ConnManAgent::Listener
```
The comment that is there now is for the method. Will document more though.

martin-ejdestig commented on src/daemon/backends/connman_technology.h:
```diff
@@ -0,0 +1,176 @@
+        SettableProperty<Glib::ustring> tethering_passphrase_;
+    };
+
+    // technology_proxy_created() is quaranteed to be called before any other method in Listener.
```
Jupp, will fix.

martin-ejdestig commented on src/daemon/backend.h:
```diff
@@ -0,0 +1,202 @@
+    // followed and call site should be fixed.
+    //
+    // WiFiAccessPoint::Event::ADDED/REMOVED_ALL are used when WiFi is enabled/disabled to limit
+    // signal emission. No WiFiAccessPoint is included in signal emission for this cases,
```
"this" -> "these"

martin-ejdestig commented on src/daemon/backends/connman_agent.cpp:
```diff
@@ -0,0 +1,336 @@
+        std::optional<Password> value_to_password(const Glib::ustring &name,
+                                                  const Glib::VariantBase &variant)
+        {
+            // FIXME: Have not tested having Glib::ustring as value here. More convenient but might
```
Forgot this FIXME... will fix. (Used FIXME for "must fix" and TODO for "punt" for now.)

martin-ejdestig commented on src/daemon/backends/connman_technology.h:
```diff
@@ -0,0 +1,176 @@
+        // signal that the value has been reverted.
+        //
+        // If setting a value in ConnMan is already pending, the new value is queued. If a value is
+        // already queued and set() is called again, the old queued value vill be replaced. This
```
"vill" -> "will"

martin-ejdestig commented on src/daemon/backends/connman_technology.h:
```diff
@@ -0,0 +1,176 @@
+        // If a value is received from ConnMan, it is updated directly if there is no pending set.
+        // If a set is pending, the value is updated when the set is finished if there is no other
+        // value queued. The received value may be due to our own set call so "property changed" is
+        // only signalled internally if the value is different, since otherwise it has already been
```
remove ","

martin-ejdestig commented on src/daemon/backends/connman_technology.h:
```diff
@@ -0,0 +1,176 @@
+
+    // technology_proxy_created() is quaranteed to be called before any other method in Listener.
+    // ConnManBackend will not consider the technology available until this has been done. This is
+    // to make sure that methods can be called and properties can be set on the technology before it
```
"can be" -> "can not be" ... this whole sentence is a bit weird... reword... do not want methods to be called before proxy has been created... will crash!

martin-ejdestig commented on src/daemon/main.cpp:
```diff
@@ -12,33 +12,46 @@
+    if (!daemon.run())
+        return EXIT_FAILURE;
 
-    return daemon.run();
```
Seems like I forgot to sync over this change (making run() return exit code instead of bool) @thenor and me did when extracting code from connectivity-manager to the template. Only noticed now when I rebased cm changes to the template. Will change to return exit code in cm as well.

martin-ejdestig commented:
> I should make a better commit message as well. Added to my TODO-list for this pr.

Have done this now. Also rebased on template that has basic build instructions in it. Will do `fixup!` commits from now on to make it easier to follow what has been changed.

TSAnsari commented on src/daemon/backends/connman_service.cpp:
```diff
@@ -0,0 +1,298 @@
+        }
+
+        template <typename T>
+        std::optional<T> value_from_variant(const Glib::VariantBase &variant,
```
This function is defined in 3 different files within anonymous namespace. Can this be made as a common function as it is being used in many files?

martin-ejdestig replied:
The log messages are unique in each one and it is only a thin wrapper around Glib::VariantBase::dynamic_cast() since it throws. But perhaps it makes sense to add the wrapping in a helper function in src/common/. Would still keep these local functions though since I think it makes sense to have different log messages (would just be changed to check return value instead of try/catch and g_warning would be kept here in local helper function).

martin-ejdestig replied:
Ideally there should be a function for this in glibmm... A version of Glib::Variant::cast_dynamic() that does not throw and returns the value directly instead of the value wrapped in a variant.

martin-ejdestig replied:
Tested adding a helper function here: https://github.com/Pelagicore/connectivity-manager/pull/1/commits/d4c4a34a171f78972c0fe08bb09d5105c2d9af5d

I do not think it is worth it. Still want the specific warning messages. So this:

```C++
        template <typename T>
        std::optional<T> value_from_variant(const Glib::VariantBase &variant,
                                            const Glib::ustring &name)
        {
            try {
                return Glib::VariantBase::cast_dynamic<Glib::Variant<T>>(variant).get();
            } catch (const std::bad_cast &) {
                g_warning("Invalid type %s for ConnMan service property \"%s\"",
                          variant.get_type_string().c_str(),
                          name.c_str());
            }
            return {};
        }
```

Becomes this:

```C++
        template <typename T>
        std::optional<T> value_from_variant(const Glib::VariantBase &variant,
                                            const Glib::ustring &name)
        {
            std::optional<T> ret = Common::variant_get<T>(variant);

            if (!ret)
                g_warning("Invalid type %s for ConnMan service property \"%s\"",
                          variant.get_type_string().c_str(),
                          name.c_str());

            return ret;
        }
```

I do not like e.g. adding a "warning string and printf arguments" to Common::variant_get()... too specific for a generic function and doing the "printf argument"-dance is not fun. Sometimes duplicating code (tm) is just the best thing to do. I think this is one of those times.

Or do you perhaps have any other suggestion for how to improve it?

TSAnsari replied:
Just wondering why not just return the value and print warning where this function is used based on the return value?

TSAnsari replied:
But I guess you don't want to print warning on every usage of this function. That would also be a type of duplication. That's why you wrapped variant_get inside value_from_variant.

TSAnsari replied:
If it is possible may be the ideal would be to modify the generator to generate code that returns typed values (As you have suggested a couple of times before). But that would be a different user story.

martin-ejdestig replied:
> But I guess you don't want to print warning on every usage of this function. That would also be a type of duplication. 

Exactly.

> But that would be a different user story.

A different story and a large one at that, I think. Since "structs" does not exist in D-Bus one would have to add custom annotations. I opened https://gitlab.freedesktop.org/dbus/dbus/issues/141 in D-Bus a couple of years ago and linked to an example (link does not work anymore though... it looks like).

martin-ejdestig replied:
But also, in this case the type in the D-Bus specification is a variant. So perhaps the "structs with names on members" part is not relevant. :)

martin-ejdestig replied:
Reverted https://github.com/Pelagicore/connectivity-manager/pull/1/commits/d4c4a34a171f78972c0fe08bb09d5105c2d9af5d and will remove it completely when squashing.

martin-ejdestig commented:
As discussed on today's standup, merging this as is now. Will make sure stories are created for the most important TODO:s.
